server/tracker.go

```diff
@@ -94,6 +117,56 @@ func (t *Tracker) count() int {
 	return len(t.connections)
 }
 
+// bannedIP returns true if the player has been banned from the server.
+func (t *Tracker) bannedIP(conn net.Conn) bool {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	ip, _, _ := net.SplitHostPort(conn.RemoteAddr().String())
+
+	banData := t.bannedIPs[ip]
+	if banData != nil && banData.score >= maxBanScore {
+		return true
+	}
+
+	return false
+}
+
+// banIP bans an IP on the server.
+func (t *Tracker) banIP(conn net.Conn) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	ip, _, _ := net.SplitHostPort(conn.RemoteAddr().String())
+
+	if _, ok := t.bannedIPs[ip]; ok {
+		t.bannedIPs[ip].score += banScoreTick
+	} else {
+		t.bannedIPs[ip] = &banData{
+			score: banScoreTick,
+		}
+	}
+
+	go t.cleanupBan(ip)
+}
+
+// cleanupBan is the decrementer on the ban score and
+// cleans up IPs that no longer need to be tracked.
+func (t *Tracker) cleanupBan(ip string) {
+	time.Sleep(banTime)
+
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	if _, ok := t.bannedIPs[ip]; ok {
+		t.bannedIPs[ip].score -= banScoreTick
+	}
+
+	if t.bannedIPs[ip].score == 0 {
```

Comment:
This might not exist? Maybe include this in the `ok` check block from above?

Reply:
Fixed.